### PR TITLE
Fix comparison of function pointers

### LINKED ISSUE #164: [Docs] Add reasons to use as-pect over typical testing frameworks for AssemblyScript

### LINKED ISSUE #166: Type declarations for as-pect
I have type declarations in https://github.com/infamous/glas/blob/master/assembly/__tests__/as-pect.d.ts. Just wondering, is that still the way to do it? Or do they get installed from NPM now?

### PR DESCRIPTION
**Pull request by willemneal**
fixes #166 
fixes #164

### CODE REVIEW

**Comment by jtenner** [comment]
Hey!

We should include a bunch of tests here for other edge cases that haven't been covered.


- [ ] `expect(funcRef).toBe(funcRef)`
- [ ] `expect(funcRef).toBe(otherFuncRef)` throws
- [ ] `expect(funcRef).not.toBe(otherFuncRef)`
- [ ] `expect(funcRef).not.toBe(funcRef)` throws
- [ ] `expect(funcRef).toStrictEqual(funcRef)`
- [ ] `expect(funcRef).toStrictEqual(otherFuncRef)` throws
- [ ] `expect(funcRef).not.toStrictEqual(otherFuncRef)`
- [ ] `expect(funcRef).not.toStrictEqual(funcRef)` throws
- [ ] `expect(funcRefArray).toContain(funcRef)`
- [ ] `expect(otherFuncRefArray).toContain(funcRef)` throws
- [ ] `expect(otherFuncRefArray).not.toContain(funcRef)`
- [ ] `expect(funcRefArray).not.toContain(funcRef)` throws

**Comment by jtenner** [comment]
On top of this, we need to add support for:

- [ ] `Actual.report<T>(funcRef)` where T is a function reference type
- [ ] `Expected.report<T>(funcRef)` where T is a function reference type
- [ ] `log<T>(funcRef)` where T is a function reference type

**Comment by jtenner** [comment]
Proposed formatting for logging function references to the console:

```
const output = `[Function ${exportIndexNumber}: ${functionDebugName}]`;
```

File: packages/assembly/assembly/__tests__/toInclude.spec.ts [hunk 1]
```diff
@@ -87,4 +93,19 @@ describe("toInclude reference arrays", () => {
...
+   */
+  it("should include a function pointer", () => {
+    expect<Array<Listener>>(eventDispatcher.events).toInclude(listener);
+    expect<Array<Listener>>(eventDispatcher.events).not.toInclude(newListener);
+  })
+
+  /**
+   * Throws if function pointer is not included.
+   */
+  throws("should include a function pointer", () => {
```
Comment (jtenner) [inline packages/assembly/assembly/__tests__/toInclude.spec.ts]: We should also add a test for `.not.toInclude(funcRef)` that throws.

This particular test seems useless, but it's saved my bacon many times!

Thank you for all this effort.

File: packages/assembly/assembly/__tests__/toIncludeEqual.spec.ts [hunk 2]
```diff
@@ -82,6 +88,22 @@ describe("toIncludeEqual reference arrays", () => {
...
+   */
+  it("should include a function pointer", () => {
+    expect<Array<Listener>>(eventDispatcher.events).toIncludeEqual(listener);
+    expect<Array<Listener>>(eventDispatcher.events).not.toIncludeEqual(newListener);
+  })
+
+  /**
+   * Throws if function pointer is not included by delegating to "toInclude".
+   */
+  throws("should include a function pointer", () => {
```
Comment (jtenner) [inline packages/assembly/assembly/__tests__/toIncludeEqual.spec.ts]: Please add `.not.toIncludeEqual` that throws as well. 

Please don't forget ending semicolons.

File: packages/assembly/assembly/__tests__/toStrictEqual.spec.ts [hunk 3]
```diff
@@ -293,4 +299,19 @@ describe("toStrictEqual", (): void => {
...
+   */
+  it("should strictly compare two function pointers", () => {
+    expect<Listener>(eventDispatcher.events[0]).toStrictEqual(listener);
+    expect<Listener>(eventDispatcher.events[0]).not.toStrictEqual(anotherListener);
+  });
+
+  /**
+   * Throws if two functions are not equal
+   */
+  throws("throws if pointers aren't equal", () => {
```
Comment (jtenner) [inline packages/assembly/assembly/__tests__/toStrictEqual.spec.ts]: Please add the `.not.toStrictEqual(listener)` throws test here too.

File: packages/assembly/assembly/internal/Expectation.ts [hunk 4]
```diff
@@ -59,8 +59,12 @@ export class Expectation<T> {
    * @param {string} message - The message that describes this assertion.
    */
   public toBe(expected: T, message: string = ""): void {
-    // assert value or reference equality
-    exactComparison<T>(this.actual, expected, this._not, message);
+    if (isFunction<T>()){
+      exactComparison<i32>(i32(this.actual), i32(expected), this._not, message);
```
Comment (jtenner) [inline packages/assembly/assembly/internal/Expectation.ts]: Clever! Will it ever be possible for a function reference pointer to be 64 bits? My guess is not.

Comment (jtenner) [reply packages/assembly/assembly/internal/Expectation.ts]: Just make sure that we add support for `Actual.report<T>()` and `Expected.report<T>()` so that the user gets a really nice output for why the test fails.

**Comment by jtenner** [comment]
Another great suggestion for this pull request might be to change the glue code:

```ts
const startFunc = wasm.__start || wasm._start; // get the start function
```

File: packages/assembly/assembly/__tests__/toIncludeEqual.spec.ts [hunk 5]
```diff
@@ -1,4 +1,10 @@
 import { Vec3 } from "./setup/Vec3";
+import { EventDispatcher, listener, anotherListener, Listener, Event } from "./setup/Event";
+
+var eventDispatcher = new EventDispatcher();
+eventDispatcher.events.push(listener);
+eventDispatcher.events.push(anotherListener);
```
Comment (9oelM) [inline packages/assembly/assembly/__tests__/toIncludeEqual.spec.ts]: Since the above three lines are used in multiple files, would it be better if we do like
```
var eventDispatcher = initializeEventDispatcher();
```
?

Comment (jtenner) [reply packages/assembly/assembly/__tests__/toIncludeEqual.spec.ts]: I'm okay with leaving it as is, because anyone reading the test can see everything that needs to be understood in a single file.

I'm also unopposed to refactoring it. @willemneal?

Comment (willemneal) [reply packages/assembly/assembly/__tests__/toIncludeEqual.spec.ts]: Fixed!

File: packages/assembly/assembly/internal/Expectation.ts (hunk 4 above)
Comment (willemneal) [reply packages/assembly/assembly/internal/Expectation.ts]: Changed this.

**Comment by willemneal** [comment]
@jtenner I'm not sure how to increase the code coverage.  I have an example log in one of the toBe tests.

**Comment by jtenner** [comment]
I can investigate later. Just need to see which lines aren't covered.

**Comment by jtenner** [comment]
Any luck with this yet? I'm curious what the function is.

**Comment by jtenner** [comment]
https://coveralls.io/builds/27536581/source?filename=packages%2Fcore%2Fsrc%2Ftest%2FTestCollector.ts#L529

Looks like we should report an Actual function, report an Expected function, and Log a function. Heh.

**Comment by jtenner** [comment]
@willemneal 

Please log a function reference here.
https://github.com/jtenner/as-pect/blob/fix-function-ref/packages/core/assembly/jest-log.ts

After this test, report a function reference that fails.
https://github.com/jtenner/as-pect/blob/fix-function-ref/packages/core/assembly/jest-pass-fail.ts#L102
```ts
expect(func).toBe(anotherFunc);
```

**Comment by willemneal** [comment]
@jtenner Thanks!  Let's hope this fixes it!

**Comment by jtenner** [comment]
Looks like that's not the case. You'll have to fix that compiler error there. Good luck @willemneal

**Comment by willemneal** [comment]
I fixed the error, but can you remind me how you update the snapshot?

**Comment by jtenner** [comment]
I believe the command is `npx jest -u`. Thank you so much for this! Snapshots make this development process quite a lot easier to assert code coverage for the glue code works, at the cost of being slightly confusing some times.

**Comment by willemneal** [comment]
[So this landed](https://github.com/AssemblyScript/assemblyscript/pull/1001) so we tomorrow we can use the nightly version to see if we can get actual function names.  But at least currently the stack trace gives some if it away.

**Comment by jtenner** [comment]
We should address #166 and bump the version among a few other things. This is turning out to be a very large pull request, and I don't mind that.

Also, docs need addressing. #169 #164 

I have been relying on you a bit much to work on this software, and I should reward you with a lot more recognition on your efforts. You deserve it!

Once that's done, we should wait for the newest version of AssemblyScript (by testing it against nightly) and release it when `0.9.0` releases.

**Comment by jtenner** [comment]
This still needs work, and coverage is currently reduced. I am voting to ignore coverage on these lines because it is a feature that is backwards compatible. WIll need to investigate a proper way to opt out of coverage here. 

cc @willemneal 

```ts
    // Getting the function name is behind an asc feature flag --exportTable, ignore coverage for this
    const func = this.wasm?.table?.get(functionPointer);
```

We should check to see if the `this.wasm?.table` property is defined first. I believe that each of these question marks end up being lines that go uncovered during tests.

**Comment by jtenner** [comment]
Also note that testing is currently being done against nightly.

**Comment by jtenner** [comment]
@willemneal We can use 

```ts
/* istanbul ignore next */ 
```

On the nullish colasce line and the `else` branch. I would imagine we should make sure the primary branch of that `if` block gets covered. Does that sound okay?

**Comment by jtenner** [comment]
@willemneal Thanks for your help. Can we review this pull request? We also might need to merge #167 into this one to remove all the excess callback signatures.

**Comment by willemneal** [comment]
@jtenner I removed what you requested and I added an ignore line for the coverage.

**Comment by jtenner** [comment]
Looks like another line needs covering. Which one is it?

**Comment by jtenner** [comment]
After that is done, we should pull it, rebase @MaxGraey 's pull request, and then go over anything that might need to be changed.

**Comment by willemneal** [comment]
@jtenner I think that's it.  The last commit is just .wat files so it's good to go!

**Comment by jtenner** [comment]
@MaxGraey thank you. Your commits were added here.
